klippa/src/colr.rs

```diff
@@ -0,0 +1,1708 @@
+                ClipList, ColorLine, ColorStop, Colr, Layer, LayerList, Paint, PaintColrGlyph,
+                PaintColrLayers, PaintComposite, PaintGlyph, PaintLinearGradient,
+                PaintRadialGradient, PaintRotate, PaintRotateAroundCenter, PaintScale,
+                PaintScaleAroundCenter, PaintScaleUniform, PaintScaleUniformAroundCenter,
+                PaintSkew, PaintSkewAroundCenter, PaintSolid, PaintSweepGradient, PaintTransform,
+                PaintTranslate, PaintVarLinearGradient, PaintVarRadialGradient, PaintVarRotate,
+                PaintVarRotateAroundCenter, PaintVarScale, PaintVarScaleAroundCenter,
+                PaintVarScaleUniform, PaintVarScaleUniformAroundCenter, PaintVarSkew,
+                PaintVarSkewAroundCenter, PaintVarSolid, PaintVarSweepGradient, PaintVarTransform,
+                PaintVarTranslate, VarAffine2x3, VarColorLine, VarColorStop,
+            },
+            variations::NO_VARIATION_INDEX,
+        },
+        FontRef, TopLevelTable,
+    },
+    types::{GlyphId, Offset24, Offset32},
+    FontBuilder,
+};
+
+// reference: subset() for COLR in Harfbuzz:
+// <https://github.com/harfbuzz/harfbuzz/blob/043980a60eb2fe93dd65b8c2f5eaa021fd8653f2/src/OT/Color/COLR/COLR.hh#L2414>
+impl Subset for Colr<'_> {
+    fn subset(
+        &self,
+        plan: &Plan,
+        _font: &FontRef,
+        s: &mut Serializer,
+        _builder: &mut FontBuilder,
+    ) -> Result<(), SubsetError> {
+        let base_glyph_list = self
+            .base_glyph_list()
+            .transpose()
+            .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?;
+        let subset_to_v0 = downgrade_to_v0(base_glyph_list.as_ref(), plan);
+
+        serialize_v0(self, plan, s, subset_to_v0)
+            .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?;
+
+        if subset_to_v0 {
+            return Ok(());
+        }
+
+        // set version to 1, format pos = 0
+        s.copy_assign(0, 1_u16);
+
+        // var_store offset pos = 30
+        if let Some(var_store) = self
+            .item_variation_store()
+            .transpose()
+            .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?
+        {
+            Offset32::serialize_subset(&var_store, s, plan, &plan.colr_varstore_inner_maps, 30)
+                .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?;
+        }
+
+        // BaseGlyphList offset pos = 14
+        Offset32::serialize_subset(&base_glyph_list.unwrap(), s, plan, (), 14)
+            .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?;
+
+        //LayerList offset pos = 18
+        if let Some(layer_list) = self
+            .layer_list()
+            .transpose()
+            .map_err(|_| SubsetError::SubsetTableError(Colr::TAG))?
+        {
+            match Offset32::serialize_subset(&layer_list, s, plan, (), 18) {
+                Ok(()) | Err(SerializeErrorFlags::SERIALIZE_ERROR_EMPTY) => (),
```

Comment:
Is the distinction between Ok(()) and `Err(ERROR_EMPTY)` ever meaningful? If it isn't, I think you should update `Offset32::serialize_subset` to check this condition and just return `Ok(())`.